char[32] data
